### PR TITLE
Run compatibility tests on CentOs

### PR DESCRIPTION
**Pull request by wolfs**
This PR configures a build on CentOS, so it uses the already built artifacts from the upstream builds to run the native platform test on CentOs 8. It replaces the artifacts built in the current build by the artifacts from an external repository via dependency substitution, so the native libraries are not rebuilt on CentOS.

### CODE REVIEW

**Comment by wolfs** [comment]
This build scan shows that the compatibility test uses the libraries from the binary repository: https://ge.gradle.org/s/popr6lh2wgk7w/dependencies?focusedDependency=WzAsNywwLFswLDcsWzBdXV0&toggled=W1swXSxbMCw3XSxbMCw3LFswXV1d

File: build.gradle [hunk 1]
```diff
@@ -15,6 +15,7 @@ versions {
 }
 
 ext.isCiServer = System.getenv().get("CI") != null
+ext.testVersionFromLocalRepository = project.hasProperty('testVersionFromLocalRepository')
```
Comment (lptr) [inline build.gradle:18]: Should we use some configuration-cache compatible way to get this property?

File: build.gradle [hunk 2]
```diff
@@ -23,6 +24,8 @@ tasks.withType(Test) {
         maxFailures = 10
         failOnPassedAfterRetry = true
     }
+    classpath = files(configurations.testRuntimeClasspath)
+    classpath.from(sourceSets.test.output)
```
Comment (lptr) [inline build.gradle:28]: I think this needs some explanation. Can you add some comments describing why we need this?
Also, why not add the two entries the same way? Say:

```groovy
classpath = files(configurations.testRuntimeClasspath, sourceSets.test.output)
```

...or perhaps:

```groovy
classpath = files()
    .from(configurations.testRuntimeClasspath)
    .from(sourceSets.test.output)
```

File: build.gradle [hunk 3]
```diff
@@ -86,9 +89,25 @@ allprojects {
 
 dependencies {
     compileOnly 'com.google.code.findbugs:jsr305:3.0.2'
+    testImplementation project(":")
```
Comment (lptr) [inline build.gradle:92]: Why is this needed?

File: build.gradle (hunk 1 above)
Comment (lptr) [reply build.gradle:18]: Also, let's add some comment here describing the purpose of this parameter.

File: build.gradle [hunk 4]
```diff
@@ -86,9 +89,25 @@ allprojects {
...
     testImplementation 'org.spockframework:spock-core:1.3-groovy-2.5'
 }
 
+if (testVersionFromLocalRepository) {
+    group = "new-group-for-root-project"
```
Comment (lptr) [inline build.gradle:97]: This looks like a considerable amount of hack that should be documented. Why do we need this?

File: build.gradle [hunk 5]
```diff
@@ -443,8 +462,10 @@ afterEvaluate {
             binary.tasks.withType(LinkSharedLibrary) { builderTask ->
                 nativeJar.into("net/rubygrapefruit/platform/$variantName") { from builderTask.linkedFile }
             }
-            project.tasks.test {
-                classpath.from nativeJar
+            if (!testVersionFromLocalRepository) {
+                project.tasks.test {
+                    classpath.from nativeJar
+                }
             }
```
Comment (lptr) [inline build.gradle:469]: Can we turn this into lazy configuration somehow?

File: build.gradle (hunk 3 above)
Comment (wolfs) [reply build.gradle:92]: We change the wiring of the test task for all runs:
Normally, the test task uses:
- output of the `main` source set
- output of the `test` source set
- testRuntimeClasspath

It is impossible to use dependency substitution for replacing the output of the main source set. So we remove the output of the main source set from the test classpath and replace it by a dependency on `project(':')` instead. That is why we need to add the dependency on project(':') here.

File: build.gradle (hunk 4 above)
Comment (wolfs) [reply build.gradle:97]: It seems like you cannot substitute a project dependency for an external dependency with the same GAV coordinates (known limitation of the engine). So I change the GAV coordinates of the root project so the dependency substitution takes effect.

File: build.gradle (hunk 5 above)
Comment (wolfs) [reply build.gradle:469]: Ah, so you want to avoid realizing the test task? I think I can do that.

File: build.gradle (hunk 1 above)
Comment (wolfs) [reply build.gradle]: Both things done.

File: build.gradle (hunk 2 above)
Comment (wolfs) [reply build.gradle]: Done.

File: build.gradle (hunk 3 above)
Comment (wolfs) [reply build.gradle:101]: I added some comments explaining what we do there.

File: build.gradle (hunk 4 above)
Comment (wolfs) [reply build.gradle:108]: I added a comment.

File: build.gradle (hunk 5 above)
Comment (wolfs) [reply build.gradle:480]: Done.

**Comment by wolfs** [comment]
@lptr PTAL!
